feat(sdk): incremental salts v2

mabsattar opened the pull request:
# Description
This PR introduces incremental proof generation functionality for MACI, allowing resumable proof generation by persisting and reusing salts between proof generation sessions. 

## Additional Notes

- **Incremental Proof Generation**: Added `--incremental` flag to CLI and SDK that enables resuming proof generation from where it left off
- **Salt Persistence**: Implemented file-based storage for proof generation salts in `.maci-salts/` directory
- **Poll-specific Storage**: Salts are stored per poll ID to support multiple concurrent polls.

### New Features
- **Salt Storage Utility**: New `saltStorage.ts` module with functions to save, load, and manage proof generation salts
- **Type Definitions**: Added `TallyVotesSalts`, `ProofGenerationSalts`, and `PollSaltsData` interfaces for type safety
- **CLI Integration**: Added `-i, --incremental` flag to the proof generation command
- **SDK Support**: Extended proof generation functions to support incremental mode.

## Technical Implementation

### Modified Components
- **Poll.tallyVotes()**: Now accepts optional salts parameter for incremental generation
- **ProofGenerator**: Enhanced to handle salt persistence and resumption
- **CLI Commands**: Updated proof generation command with incremental option

### File Changes
- Added `.maci-salts/` to `.gitignore` to prevent accidental commit of salts
- Updated Poll.ts to support optional salts in tallyVotes method
- Enhanced ProofGenerator with salt management logic
- Added comprehensive type definitions for salt interfaces

## Related issue(s)

- Supersedes #2783 (previous incomplete implementation of incremental proof generation)
- Builds upon foundation from #2337 (earlier proof generation improvements)

## Usage

```bash
# Generate proofs incrementally (resumes if possible)
node build/cli/index.js genProofs -i -o <pollId> [other options]

# Generate proofs from scratch (default behavior)
node build/cli/index.js genProofs -o <pollId> [other options]
```

## Testing Checklist
- [x] All existing tests pass
- [x] Code builds without errors (`pnpm build`)
- [x] CLI flag appears in help: `maci generateProofs --help`
- [x] TypeScript types are correct
- [x] `.gitignore` updated

## Confirmation
- [yes ] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [ yes] I ran and verified that all tests pass according to MACI's [testing guide](https://maci.pse.dev/docs/guides/testing).

mabsattar committed:
feat(proofs) added types and a file storage utility to persist salts for incremental proof generation

mabsattar committed:
Modify Poll.tallyVotes() to accept optional salts parameter

mabsattar committed:
Update ProofGenerator to save/load salts automatically

mabsattar committed:
Add TallyVotesSalts, ProofGenerationSalts, PollSaltsData interfaces

mabsattar committed:
fix: update ProofGenerator instantiation with required pollId parameter

mabsattar committed:
feat: add incremental flag to CLI and SDK for resumable proof generation

mabsattar committed:
chore: add .maci-salts to .gitignore

mabsattar committed:
Merge branch 'privacy-ethereum:main' into feat/incremental-salts-v2

NicoSerranoP commented on packages/contracts/tasks/helpers/ProofGenerator.ts:
```diff
@@ -69,6 +71,10 @@ export class ProofGenerator {
    */
   private rapidsnark?: string;
 
+  private incremental: boolean;
```
Please add TS doc comments for each new parameter

NicoSerranoP commented on packages/contracts/tasks/helpers/types.ts:
```diff
@@ -222,6 +222,10 @@ export interface IProofGeneratorParams {
+
+  incremental?: boolean;
+
+  pollId: string;
```
Add TS-docs

NicoSerranoP commented on packages/contracts/tasks/helpers/types.ts:
```diff
@@ -900,3 +904,26 @@ export interface IDeployContractWithLinkedLibrariesParams {
+  newResultsRootSalt: string;
+  newSpentVoiceCreditSubtotalSalt: string;
+  newPerVoteOptionSpentVoiceCreditsRootSalt: string;
+  tallyBatchNum?: number;
```
always use the complete name. Rename `tallyBatchNum` to `tallyBatchNumber`

mabsattar committed:
Merge branch 'privacy-ethereum:main' into feat/incremental-salts-v2

mabsattar committed:
feat(contracts): add ts doc comments replace log msgs and rename tallyBatchNum to tallyBatchNumber

mabsattar committed:
feat(contracts): add ts doc comments and fix assignment

mabsattar committed:
feat(sdk): fix broken js doc comment

mabsattar committed:
async file writing

mabsattar committed:
rename saltStorage.ts to SaltStorage.ts

mabsattar committed:
refactor(contracts): add await saveSalts, fix import, and update logging

mabsattar committed:
fix(contracts): extract salts directly from circuit inputs

mabsattar committed:
fix(contracts): add TS doc comments for incremental proof generation types

mabsattar committed:
fix(core): remove redundant salts cast in Poll.tallyVotes()

mabsattar committed:
fix(sdk): wire incremental flag through generateProofs and fix formatting

mabsattar committed:
fix: address PR feedback and linting issues

mabsattar committed:
fix(sdk): add JSDoc for incremental flag and clean up comments

mabsattar commented:
hey @NicoSerranoP I've pushed the updates as per your feedback.
- fixed the jsdoc comments and formatting issues.
- switched `saltStorage` to use async/await
- regarding the salt saving logic in `Poll.tallyVotes()`: I kept the actual file I/O inside `ProofGenerator.ts` to avoid importing fs into packages/core, which would break browser compatibility. `tallyVotes` handles the logic, and `ProofGenerator` handles the persistence
